bmap-rs/src/main.rs

```diff
@@ -74,53 +83,150 @@ impl SeekForward for Decoder {
         Some("gz") => {
             let gz = GzDecoder::new(f);
-            Ok(Decoder::new(Discarder::new(gz)))
+            Ok(Input::Local(Decoder::new(Discarder::new(gz))))
         }
-        _ => Ok(Decoder::new(f)),
+        _ => Ok(Input::Local(Decoder::new(f))),
     }
 }
 
-fn copy(c: Copy) -> Result<()> {
-    if !c.image.exists() {
-        bail!("Image file doesn't exist")
-    }
+async fn copy(c: Copy) -> Result<()> {
+    let url = c
+        .image
+        .to_str()
+        .expect("Fail to convert remote path to str")
+        .parse::<Uri>()
+        .expect("Fail to convert remote path to url");
+    let input_type = match url.scheme() {
+        Some(_) => setup_remote_input(url.clone(), &c.image).await?,
+        None => {
+            if !c.image.exists() {
+                bail!("Image file doesn't exist")
+            }
+            setup_local_input(&c.image)?
+        }
+    };
+
+    let bmap = match input_type {
+        Input::Local(_) => find_bmap(&c.image).ok_or_else(|| anyhow!("Couldn't find bmap file"))?,
+        Input::Remote(_) => {
+            let mut url_bmap = c.image.clone();
+            url_bmap.set_extension("bmap");
+            url_bmap
+        }
+    };
 
-    let bmap = find_bmap(&c.image).ok_or_else(|| anyhow!("Couldn't find bmap file"))?;
     println!("Found bmap file: {}", bmap.display());
 
-    let mut b = File::open(&bmap).context("Failed to open bmap file")?;
-    let mut xml = String::new();
-    b.read_to_string(&mut xml)?;
-
-    let bmap = Bmap::from_xml(&xml)?;
-    let mut output = std::fs::OpenOptions::new()
-        .write(true)
-        .create(true)
-        .open(c.dest)?;
+    let bmap = match input_type {
+        Input::Local(_) => {
+            let mut b = File::open(&bmap).context("Failed to open bmap file")?;
+            let mut xml = String::new();
+            b.read_to_string(&mut xml)?;
+            Bmap::from_xml(&xml)?
+        }
 
-    if output.metadata()?.is_file() {
-        ftruncate(output.as_raw_fd(), bmap.image_size() as i64)
-            .context("Failed to truncate file")?;
+        Input::Remote(_) => {
+            let url_bmap = bmap
+                .to_str()
+                .expect("Fail to convert remote path to str")
+                .parse::<Uri>()
+                .expect("Fail to convert remote path to url");
+            let mut bmap_res = match url_bmap.scheme_str() {
+                Some("https") => fetch_url_https(url_bmap).await?,
+                Some("http") => fetch_url_http(url_bmap).await?,
+                _ => {
+                    bail!("url is not http or https")
+                }
+            };
+            let mut c = Cursor::new(Vec::new());
+            while let Some(next) = bmap_res.data().await {
+                let chunk = next?;
+                c.write_all(&chunk)?;
+            }
+            c.set_position(0);
+            let mut xml = String::new();
+            c.read_to_string(&mut xml)?;
```

Comment:
Ignoring the fact you onjly need to solve this due to using Hyper; You don't want to use a cursor here or do this loop by hand, hyper has its `hyper::body::to_bytes` helper that collects everything already;  followed by `std::str::from_utf8`